### PR TITLE
Upgrade auth stack

### PR DESCRIPTION
**Pull request by thomas11**
This PR upgrades the authentication dependencies such as azure-go-helpers to the latest version, which is a jump of many versions. It also upgrades some other dependencies like the pulumi sdk.

The authentication method is still the deprecated ADAL, not MSAL, which has an additional requirement of specifying the token audience which is tricky for us; see #1942.

### CODE REVIEW

File: examples/go-clientconfig/go.mod [hunk 1]
```diff
@@ -0,0 +1,65 @@
+module main
```
Comment (lblackstone) [inline examples/go-clientconfig/go.mod:1]: It's not generally necessary to include a `go.mod` in the test folder. The integration test framework will automatically generate one.

File: provider/pkg/provider/provider.go [hunk 2]
```diff
@@ -213,10 +216,11 @@ func (k *azureNativeProvider) Invoke(ctx context.Context, req *rpc.InvokeRequest
 		}
 		objectId := ""
 		if auth.GetAuthenticatedObjectID != nil {
-			objectId, err = auth.GetAuthenticatedObjectID(ctx)
+			objectIdPtr, err := auth.GetAuthenticatedObjectID(ctx)
 			if err != nil {
 				return nil, fmt.Errorf("getting authenticated object ID: %w", err)
 			}
+			objectId = *objectIdPtr
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:223]: Is the pointer guaranteed to be non-nil if err is nil?

File: provider/pkg/provider/provider.go [hunk 3]
```diff
@@ -176,7 +176,10 @@ func (k *azureNativeProvider) Configure(ctx context.Context,
 	}
 	k.environment = env
 
-	tokenAuth, bearerAuth, err := k.getAuthorizers(authConfig)
+	// The ctx Context given by gRPC is request-scoped and will be canceled after this request. We
+	// need the authorizers to function across requests.
+	authCtx := context.Background()
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:181]: It looks like you might be able to create the context inside of the `getAuthorizers` method and drop the method signature change.

File: examples/go-clientconfig/go.mod (hunk 1 above)
Comment (thomas11) [reply examples/go-clientconfig/go.mod:1]: All of the `go-*` tests in this repo have it. When I remove it from my test case, it fails with

```
github.com/pulumi/pulumi-azure-native/sdk/go/azure/authorization: github.com/pulumi/pulumi-azure-native/sdk/go@v0.0.0-00010101000000-000000000000: replacement directory /Users/tkappler/go/src/github.com/pulumi/pulumi-azure-native/sdk/go does not exist
```

File: provider/pkg/provider/provider.go (hunk 3 above)
Comment (thomas11) [reply provider/pkg/provider/provider.go]: Good catch, that's better factoring.

File: provider/pkg/provider/provider.go (hunk 2 above)
Comment (thomas11) [reply provider/pkg/provider/provider.go:223]: Based on my reading of the source it is, but it's not officially guaranteed, so I added a check.

**Review by danielrbradley — APPROVED** [review]

**Comment by danielrbradley** [comment]
As a general question - I'm assuming our test coverage is quite low for the different authentication flows given all our tests run using the same credentials? Do you have any thoughts on what would be involved in testing some of our other auth flows?

**Pull request merged**